### PR TITLE
Convert project into TypeScript

### LINKED ISSUE #122: Convert project into TypeScript
Historically, there have been [a good number of TypeScript issues](https://github.com/kolodny/immutability-helper/issues?q=is%3Aissue+typescript+is%3Aclosed) filed against this repository. If the source code were in TypeScript then the definitions would be generated, solving the most common TypeScript issues that might come up. Furthermore, any maintenance to the source code would remain in sync with its definitions automatically, as they are generated.

### PR DESCRIPTION
**Pull request by jednano**
Fixes #122 
Fixes #117 
Fixes #127 

Based on and fixes #126 

I've separated into 4 commits for an easier review.

Concerning the last commit, [Microsoft says the following](https://github.com/Microsoft/dtslint#setup):

> If you are writing the library in TypeScript, don't use `dtslint`. Use [`--declaration`](http://www.typescriptlang.org/docs/handbook/compiler-options.html) to have type definitions generated for you.

### CODE REVIEW

**Comment by jednano** [comment]
@kolodny this PR is good to go. Let me know how you feel about it!

File: index.ts
Comment (jednano) [inline index.ts]: This is the only ignore line I added and I have no idea why I had to. @kolodny do you know which test this relates to? I didn't add or remove any tests.

**Comment by kolodny** [comment]
This is amazing, thanks for this effort! @andreiglingeanu what are your thoughts on this? I'm not a ts expert by any means so I may need some time to review this.

**Review by andreiglingeanu — APPROVED** [review]
Just played a bit with this PR for a bit, looks very clean, definitely a step forward!

This definitely is not a breaking change for existing non-TS users while providing a better TS workflow. I think this is good to go, if @kolodny is fine with it

File: test.ts
Comment (jednano) [inline test.ts:1]: `use strict` combined with `Object.freeze` will throw an error if you attempt to mutate an object, so adding this `use strict` to the top here enables that functionality for the rest of the tests in this file.

File: test.ts
Comment (jednano) [inline test.ts]: Removed this `expect(obj).toEqual(xxx)` paradigm from the entire test file, as it's not actually testing what we want to test (i.e., whether the object was mutated).

File: test.ts
Comment (jednano) [inline test.ts]: When `use strict` is enabled it's not happy with `null` as the lexical `this` context.

File: index.ts
Comment (kolodny) [reply index.ts]: I'm not sure, this is ignoring this line, so we should have better test coverage to try to catch this line. A lot of folks are into forcing 100% coverage especially for library code but I'm OK with it a little lower. My preference would be to remove since this is something we could cover unlike the polyfilled `Object.assign` or some other polyfill code. Up to you either way

File: typings-test.ts
Comment (kolodny) [inline typings-test.ts:1]: I noticed running `npm run lint` produces 

```
> immutability-helper@2.8.1 lint /Users/mkolodny/projects/immutability-helper
> tslint -p tsconfig.test.json

No valid rules have been specified for TypeScript files
```

Is there something we should be doing instead?

**Review by kolodny — APPROVED** [review]
After running `npm run build` There are some more changes in `index.d.ts`, should those also be included in the commit?

I'm going to approve these changes. Once again I want to thank you for taking this on and for the work put into it

File: tsconfig.json
Comment (kolodny) [inline tsconfig.json:8]: Why was this changed to false?

Comment (jednano) [reply tsconfig.json:8]: So I didn't have to scatter `: any` everywhere in the file. I'm thinking baby steps. If this PR gets in I can start working on defining more of these complex types, because there are so many that are not defined yet.

File: typings-test.ts
Comment (jednano) [reply typings-test.ts:1]: Looks like [a TSLint issue](https://github.com/palantir/tslint/issues/3735), but it's resolved in 67360db. Thanks for pointing this out!

**Comment by jednano** [comment]
@kolodny the reason your `d.ts` file changed is because I did it on Windows and you did it on non-Windows. I've fixed that in 16c3ad0.

I think we can `/* istanbul ignore next */` for now and address any of those in future commits. It's really outside of the scope of this PR.

Thanks for the ❤️ words and for reviewing this PR. I hope (time permitting) to submit some more with more complex types defined in the future.

**Comment by jednano** [comment]
Who typically merges the PRs after approval?

**Comment by kolodny** [comment]
I'm going to take one more pass, try to add tests that it installs and runs properly on a non ts workflow (feel free to do that if you like) merge it in and push a release. Thanks for your patience

**Comment by kolodny** [comment]
@jedmao Is there a sane way to still allow folks to do  
`const update = require('immutability-helper');` and not need to do  
`const update = require('immutability-helper').default;`?

File: index.ts
Comment (jednano) [inline index.ts]: This had to be moved inside the new namespace.

File: index.ts
Comment (jednano) [inline index.ts]: I don't ❤️  that I have to create this `IUpdate` interface, but it does simplify things for the exported types.

File: index.ts
Comment (jednano) [inline index.ts]: Since this module is exporting a `default` function, but also TypeScript types that compile into nothing (e.g., `CustomCommands` and `Spec`), we use a TypeScript namespace to declare the group of members for export. It's a bit wonky, but it works.

File: index.ts
Comment (jednano) [inline index.ts]: TypeScript allows you to merge namespaces with functions, so long as they share the same name. That's what I'm doing here. Unfortunately, I can't do `const update = update.newContext()` and achieve the same result. Wish I could!

File: index.ts
Comment (jednano) [inline index.ts]: This is the first place where the new `IUpdate` interface proves valuable.

File: index.ts
Comment (jednano) [inline index.ts]: All these types are just indented one level, because they had to move inside the namespace.

File: index.ts
Comment (jednano) [inline index.ts]: These next few lines are not lovely, but I can tell you that I've been trying to do exactly this with TypeScript for years now and never figured it out until just now. Since my new `update` function is really just a pass-thru to the `defaultContext`, we need to assign the same members to it with `Object.assign`. TypeScript is still unaware of the structure of the resulting object, but that's OK because I just export it as `IUpdate` and we're good to go.

File: test.ts
Comment (jednano) [inline test.ts]: You wanted to test that this works with non-TypeScript libraries. I think this require statement and the tests that follow prove that it does and that it's 100% in sync w/ the TypeScript imports.

https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require

File: test.ts
Comment (jednano) [inline test.ts]: I wrote these tests after testing the following in Bash:
```bash
~/.../GitHub/immutability-helper (typescript)$node
> var update = require('.')
undefined
> update
{ [Function: update]
  newContext: [Function: newContext],
  extend: [Function],
  isEquals: [Function],
  default: [Circular] }
>
```
See that `update.default` is a circular reference, which is what we want.

**Review by jednano — COMMENTED** [review]
@kolodny no. There is no "sane" way to do it. Not from what I've seen. If you want to have both import systems then it's going to be painful without something like [Rollup](https://rollupjs.org/guide/en) doing it for you.

https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html#commonjs-interop

That said, we could talk about doing what @basarat is saying in the link above and remove the whole concept of a default export. We can still `export = update`.

**Comment by jednano** [comment]
Build is still failing on older versions of Node. I have some ideas to clean up the code too. Stay tuned.

**Comment by jednano** [comment]
Ugh! Nothing "clean" is working. Maybe we can just do it the TypeScript way and start using Rollup?

**Comment by kolodny** [comment]
I found a pretty nasty solution that fixes. The basic idea is to do something like

```js
const defaultExport = newContext();
declare var module: any;
// @ts-ignore
declare var exports: any;
module.exports = exports = defaultExport;
export default defaultExport;
```

But at this point I'm not sure if it's worth doing just for the commonjs folks. What are your thoughts? Curious as to what the rest of the community as well as @andreiglingeanu thoughts on this is as well. Is this worth bothering with or should be just require people to use  
`const update = require('immutability-helper').default;` from now on? We could just jump a major version and ship it at fbf196b

**Comment by jednano** [comment]
I want some more time to play with this. I don't love that anyone should have to add `.default` to an imported module. I fixed this with PostCSS, but that library is a bit different because it really does export a default function and not a new variable like this library does with `new Context()`.

That said, I'm all for breaking the version and actually changing the default export to a function/namespace and getting rid of this default new Context idea, but it requires a bit more thought.

File: README.md
Comment (jednano) [inline README.md]: #### Breaking change
No more default exports, so every import is in the format `import { x, y, z } from 'immutability-helper';`

File: index.ts
Comment (jednano) [inline index.ts]: A new `defaultContext` is exported here, which was previously the `default` export.

File: index.ts
Comment (jednano) [inline index.ts]: This exported `update` function is similar to the previous `default` export except that this one is an actual `Context` class method and has no `.extend` or `.isEquals` properties attached to it.

File: index.ts
Comment (jednano) [inline index.ts]: `isEquals`, `extend` and `update` are all `Context` class methods now.

File: package.json
Comment (jednano) [inline package.json]: Mocha was giving me so much trouble, failing on Travis-CI builds for Node.js v6 and v8. I finally threw in the towel and switched to Jest, which seemed to solve all my problems!

File: test-js.js
Comment (jednano) [inline test-js.js]: Since we're introducing breaking changes anyway with the exported members, I figured this smoke test doesn't hold a lot of value any more. Correct me if I'm wrong.

**Review by jednano — COMMENTED** [review]
I'm pretty happy with where this landed, but it does introduce breaking changes. @kolodny I'd like to hear your take on this. I do feel like this is more clean and less hacky than what was attempted before.

File: README.md
Comment (andreiglingeanu) [reply README.md]: What if we keep the default export and re-export it as `update`?
Users can always do `import update, { defaultContext as context } from 'immutability-helper'` if they want to.

I agree that having `require('immutability-helper').default` in cjs is a bit nasty but breaking ES modules seems a bit nasty at first glance...

**Comment by andreiglingeanu** [comment]
The work here looks so awesome! 🙂 If there is needed more feedback/testing/anything before merging this -- I'm up for it!

File: README.md
Comment (jednano) [reply README.md]: Give me a moment to flush this out.

Comment (jednano) [reply README.md]: See 4fc6303

1. I am no longer exporting `defaultContext`, because I am now exporting its `update` and `extend` methods, bound to the default context (as the library already does).
1. I removed support for `update.extend` for two reasons:
   - Two ways of doing the same thing is generally confusing.
   - It's hard, if not impossible to add properties to a method. I can do it with an interface and have `Context implements IUpdate`, but it doesn't compile that way.
1. I removed any reference to `require('immutability-helper').default` in the README because we shouldn't advertise this as a golden path. Let's call it an undocumented feature.

**Comment by kolodny** [comment]
BTW putting

```js
// @ts-ignore
exports.default.default = module.exports = exports.default;
```

also works

File: README.md
Comment (jednano) [inline README.md]: Backwards compatibility met!

File: index.ts
Comment (jednano) [inline index.ts]: This will support current implementations of `update.extend` for now (deprecated), but it's now an undocumented path and unsupported by the TypeScript definitions.

**Review by jednano — COMMENTED** [review]
~We have 100% backwards compatibility with these updates! Thanks for the help, guys!~

**Edit:** there are some breaking changes called out in https://github.com/kolodny/immutability-helper/pull/123#issuecomment-445385167.

**Comment by jednano** [comment]
OK I don't know why the smoke test isn't working on Travis-CI. Works fine on my Windows machine. I have to work on something else for now. Any help would be appreciated to get this build working.

**Comment by jednano** [comment]
Everything works!

**Comment by jednano** [comment]
# Changelog (draft)

## 3.0.0 (presumably)

- **New:** TypeScript definitions are now generated from TypeScript source. meaning they will always be 100% in sync w/ the source code.
- **New:** exported `Context` class.
   - **Deprecated:** stop using `const myUpdate = newContext().update` and start using `const myContext = new Context(); myContext.update(...)`.
- **Deprecated:** stop using `update.extend` and start using `import { extend } from 'immutability-helper'` or `const { extend } = new Context()`.

**Comment by jednano** [comment]
If we're not cool with the breaking changes I can revert 68620c2 and export a `newContext`, but I don't love that there will be 2 ways to do these things (e.g., `new Context()` vs. `newContext()`). Granted, I _could_ remove the `Context` export if we want to keep it hidden.

**Comment by jednano** [comment]
@kolodny @andreiglingeanu I've restructured this PR into relevant commits, easing your fears by proving that the build and smoke test pass both before and after updating the original test file and source. Let me know if you want me to fix anything. Thanks!

File: README.md
Comment (andreiglingeanu) [reply README.md]: Sounds great!

**Comment by andreiglingeanu** [comment]
All good and green on my side! Everything runs smoothly as far as I tested, this work is so good! @kolodny I think this can be merged when you feel like ready for it.

@jedmao this PR includes the stuff from https://github.com/kolodny/immutability-helper/pull/126, is that correct?

**Comment by jednano** [comment]
Yes, but I switched it from Packagephobia to Bundlephobia. Once this PR is merged it will automatically close #126, #122 and #117.

**Comment by jednano** [comment]
This PR no longer has any breaking changes, so I think the changelog can look like this:


# Changelog

## 2.9.0

- **New:** TypeScript definitions are now generated from TypeScript source. meaning they will always be 100% in sync w/ the source code.
- **New:** exported `Context` class.
   - **Deprecated:** stop using `const myUpdate = newContext().update` and start using `const myContext = new Context(); myContext.update(...)`.
- **Deprecated:** stop using `update.extend` and start using `import { extend } from 'immutability-helper'` or `const { extend } = new Context()`.

**Comment by jednano** [comment]
@kolodny are you waiting on something? Let me know.

**Comment by kolodny** [comment]
I wanted to do a final pass - done now. I'm going to merge this in and then publish a release with the release notes you wrote.

I'm also going to give you commit access to this repo.

I just wanted to say again: Thanks for this PR and the followup to land this PR. This is amazing work and will greatly benefit the community.

**Pull request merged**
